From 4ee0986f59a7cea510dbb1edcb46d54c259b0f40 Mon Sep 17 00:00:00 2001
From: Patrick Williams <patrick@stwcx.xyz>
Date: Wed, 5 Mar 2025 16:12:45 -0500
Subject: [PATCH] CPER: add generic event for CPER errors

We do not have a high-level design for CPER errors from libcper and/or
SMCs, but these devices are able to create CPER events.  Define simple
generic CPER event definitions as a placeholder so that we can start
utilizing them.  As time goes on, I anticipate that there may be some
generation of events from libcper that gives finer granularity of the
individual CPER events and these may become deprecated.

Signed-off-by: Patrick Williams <patrick@stwcx.xyz>
Change-Id: Icffe6aa892cc57c4f0f9e0c326e53d2ba92e27cf
---
 .../openbmc_project/State/CPER/meson.build    | 26 +++++++++++
 gen/xyz/openbmc_project/State/meson.build     | 45 +++++++++++++++++++
 .../openbmc_project/State/CPER.events.yaml    | 34 ++++++++++++++
 3 files changed, 105 insertions(+)
 create mode 100644 gen/xyz/openbmc_project/State/CPER/meson.build
 create mode 100644 yaml/xyz/openbmc_project/State/CPER.events.yaml

diff --git a/gen/xyz/openbmc_project/State/CPER/meson.build b/gen/xyz/openbmc_project/State/CPER/meson.build
new file mode 100644
index 0000000..ef93af1
--- /dev/null
+++ b/gen/xyz/openbmc_project/State/CPER/meson.build
@@ -0,0 +1,26 @@
+# Generated file; do not modify.
+
+sdbusplus_current_path = 'xyz/openbmc_project/State/CPER'
+
+generated_sources += custom_target(
+    'xyz/openbmc_project/State/CPER__cpp'.underscorify(),
+    input: ['../../../../../yaml/xyz/openbmc_project/State/CPER.events.yaml'],
+    output: ['event.cpp', 'event.hpp'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'cpp',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../../yaml',
+        'xyz/openbmc_project/State/CPER',
+    ],
+    install: should_generate_cpp,
+    install_dir: [false, get_option('includedir') / sdbusplus_current_path],
+    build_by_default: should_generate_cpp,
+)
+
diff --git a/gen/xyz/openbmc_project/State/meson.build b/gen/xyz/openbmc_project/State/meson.build
index 1d3b4d4..25c93b1 100644
--- a/gen/xyz/openbmc_project/State/meson.build
+++ b/gen/xyz/openbmc_project/State/meson.build
@@ -2,6 +2,7 @@
 subdir('BMC')
 subdir('BMCRedundancy')
 subdir('Boot')
+subdir('CPER')
 subdir('Cable')
 subdir('Chassis')
 subdir('Decorator')
@@ -97,6 +98,50 @@ generated_markdown += custom_target(
     build_by_default: should_generate_markdown,
 )
 
+generated_markdown += custom_target(
+    'xyz/openbmc_project/State/CPER__markdown'.underscorify(),
+    input: ['../../../../yaml/xyz/openbmc_project/State/CPER.events.yaml'],
+    output: ['CPER.md'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'markdown',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../yaml',
+        'xyz/openbmc_project/State/CPER',
+    ],
+    install: should_generate_markdown,
+    install_dir: [inst_markdown_dir / sdbusplus_current_path],
+    build_by_default: should_generate_markdown,
+)
+
+generated_registry += custom_target(
+    'xyz/openbmc_project/State/CPER__registry'.underscorify(),
+    input: ['../../../../yaml/xyz/openbmc_project/State/CPER.events.yaml'],
+    output: ['CPER.json'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'registry',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../yaml',
+        'xyz/openbmc_project/State/CPER',
+    ],
+    install: should_generate_registry,
+    install_dir: [inst_registry_dir / sdbusplus_current_path],
+    build_by_default: should_generate_registry,
+)
+
 generated_markdown += custom_target(
     'xyz/openbmc_project/State/Cable__markdown'.underscorify(),
     input: ['../../../../yaml/xyz/openbmc_project/State/Cable.events.yaml'],
diff --git a/yaml/xyz/openbmc_project/State/CPER.events.yaml b/yaml/xyz/openbmc_project/State/CPER.events.yaml
new file mode 100644
index 0000000..3d37809
--- /dev/null
+++ b/yaml/xyz/openbmc_project/State/CPER.events.yaml
@@ -0,0 +1,34 @@
+version: 1.0.0
+
+errors:
+    - name: GenericCPERWarning
+      severity: warning
+      metadata:
+          - name: Source
+            type: object_path
+            primary: true
+            description: The name or identifier of the CPER source.
+          - name: CPER
+            type: string
+            primary: true
+            description: The CPER JSON of the event.
+      en:
+          description: A generic CPER warning has been detected.
+          message: A generic CPER warning has been detected on {Source}.
+      errno: EIO
+
+    - name: GenericCPERFault
+      severity: critical
+      metadata:
+          - name: Source
+            type: object_path
+            primary: true
+            description: The name or identifier of the CPER source.
+          - name: CPER
+            type: string
+            primary: true
+            description: The CPER JSON of the event.
+      en:
+          description: A generic CPER fault has been detected.
+          message: A generic CPER fault has been detected on {Source}.
+      errno: EIO
-- 
2.47.3

